pipeline_dp/budget_accounting.py

```diff
@@ -104,6 +109,64 @@ def request_budget(
     def compute_budgets(self):
         pass
 
+    def scope(self, weight: float):
+        """
+        Defines a scope for DP operations that should consume no more than "weight" proportion of the budget
+        of the parent scope. The accountant will automatically scale the budgets of all sub-operations accordingly.
+
+        Example usage:
+          with accountant.scope(weight = 0.5):
+             ... some code that consumes DP budget ...
+
+        :param weight: budget weight of all operations made within this scope as compared to.
+        :return: the scope that should be used in a "with" block enclosing the operations consuming the budget.
+        """
+        return BudgetAccountantScope(self, weight)
+
+    def _register_mechanism(self, mechanism: MechanismSpecInternal):
+        """Registers this mechanism for the future normalisation."""
+
+        # Register in the global list of mechanisms
+        self._mechanisms.append(mechanism)
+
+        # Register in all of the current scopes
+        for scope in self._scopes:
+            scope.mechanisms.append(mechanism)
+
+        return mechanism
+
+    def _enter_scope(self, scope):
+        self._scopes.append(scope)
+
+    def _exit_scope(self):
+        self._scopes.pop()
+
+
+
+
+@dataclass
+class BudgetAccountantScope:
+    def __init__(self, accountant, weight):
+        self.weight = weight
+        self.accountant = accountant
+        self.mechanisms = []
+
+    def __enter__(self):
+        self.accountant._enter_scope(self)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.accountant._exit_scope()
+        self._normalise_mechanism_weights()
+
+    def _normalise_mechanism_weights(self):
+        """Normalise all mechanism weights so that they sum up to no more than the weight of the current scope."""
```

Comment:
It looks that it normilises such that sum is equal self.weigh, please remove "no more"